### PR TITLE
Add sharding options to test command line

### LINKED ISSUE #71225: Add option `--total-shards` to flutter test command
# Description
I can see in test package docs https://pub.dev/packages/test#sharding-tests that there is one param to shard the tests:

```bash
pub run test --total-shards 3 --shard-index 0 path/to/test.dart
pub run test --total-shards 3 --shard-index 1 path/to/test.dart
pub run test --total-shards 3 --shard-index 2 path/to/test.dart
```

And when I try to run this in my project I get:
```bash
Could not find package "test". Did you forget to add a dependency?
pub finished with exit code 65
```

But when I try to run the test command from flutter I can see this:
```bash
$ flutter test  --total-shards 3 --shard-index 0  path/to/test.dart
Could not find an option named "total-shards".

Run 'flutter -h' (or 'flutter <command> -h') for available flutter commands and options.
```

and when I run `flutter -h`:

```bash
$ flutter test -h
Run Flutter unit tests for the current project.

Global options:
-h, --help                  Print this usage information.
-v, --verbose               Noisy logging, including all shell commands executed.
                            If used with --help, shows hidden options.
-d, --device-id             Target device id or name (prefixes allowed).
    --version               Reports the version of this tool.
    --suppress-analytics    Suppress analytics reporting when this command runs.

Usage: flutter test [arguments]
-h, --help                            Print this usage information.
    --[no-]pub                        Whether to run "flutter pub get" before executing this command.
                                      (defaults to on)
    --[no-]null-assertions            Perform additional null assertions on the boundaries of migrated and unmigrated code. This
                                      setting is not currently supported on desktop devices.
    --[no-]track-widget-creation      Track widget creation locations. This enables features such as the widget inspector. This
                                      parameter is only functional in debug mode (i.e. when compiling JIT, not AOT).
                                      (defaults to on)
    --name=<regexp>                   A regular expression matching substrings of the names of tests to run.
    --plain-name=<substring>          A plain-text substring of the names of tests to run.
-t, --tags                            Run only tests associated with tags
-x, --exclude-tags                    Run only tests WITHOUT given tags
    --start-paused                    Start in a paused mode and wait for a debugger to connect.
                                      You must specify a single test file to run, explicitly.
                                      Instructions for connecting with a debugger are printed to the console once the test has started.
    --coverage                        Whether to collect coverage information.
    --merge-coverage                  Whether to merge coverage data with "coverage/lcov.base.info".
                                      Implies collecting coverage data. (Requires lcov)
    --coverage-path                   Where to store coverage information (if coverage is enabled).
                                      (defaults to "coverage/lcov.info")
    --update-goldens                  Whether matchesGoldenFile() calls within your test methods should update the golden files rather
                                      than test for an existing match.
-j, --concurrency=<jobs>              The number of concurrent test processes to run.
                                      (defaults to "10")
    --[no-]test-assets                Whether to build the assets bundle for testing.
                                      Consider using --no-test-assets if assets are not required.
                                      (defaults to on)
    --platform                        The platform to run the unit tests on. Defaults to "tester".
                                      [tester (default), chrome]
    --test-randomize-ordering-seed    The seed to randomize the execution order of test cases.
                                      Must be a 32bit unsigned integer or "random".
                                      If "random", pick a random seed to use.
                                      If not passed, do not randomize test case execution order.
```

I do see `--test-randomize-ordering-seed` that is also in the test docs https://pub.dev/packages/test#shuffling-tests.

It looks to me that this feature is missing from flutter itself or maybe I am not looking where I should.

# Env
```
flutter: 1.22.3

dev_dependencies:
  flutter_test:
    sdk: flutter
```

### PR DESCRIPTION
**Pull request by danielgomezrico**
The idea is to be able to run tests in parallel and this feature is already on the test package, but flutter CLI does not recognize them.

- It will add two parameters:
```
--total-shards 3 
--shard-index 0
````
   - I added some validations:
      - `--total-shards`: should be positive and if set then `--shard-index` must exist also
      - `--shard-index`: should be more than -1 and if set then `--shard-index` must exist also
- Update docs on -h adding this two options:
```
    --total-shards                    Tests can be sharded with the "--total-shards" and "--shard-index" arguments, allowing you to split up your test suites and run them separately.
    --shard-index                     Tests can be sharded with the "--total-shards" and "--shard-index" arguments, allowing you to split up your test suites and run them separately.
```

This PR will fix https://github.com/flutter/flutter/issues/71225

# Current

```bash
$ flutter test  --total-shards 3 --shard-index 0  path/to/test.dart
Could not find an option named "total-shards".
```

# Expected

```bash
$ flutter test  --total-shards 3 --shard-index 0  path/to/test.dart
... tests run fine
```

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by danielgomezrico** [comment]
@christopherfujino I tested locally and fixed tests on https://github.com/flutter/flutter/blob/0e377069a1bbce7ad245748aec0fdb27ba5e9c06/packages/flutter_tools/test/commands.shard/hermetic/test_test.dart#L39 but Im not sure why some tasks are failing :( can I get some advice here?

**Comment by danielgomezrico** [comment]
Oh they are passing now after a rebase from master :)

**Review by otuzoglu — APPROVED** [review]

File: packages/flutter_tools/test/commands.shard/hermetic/test_test.dart
Comment (christopherfujino) [inline packages/flutter_tools/test/commands.shard/hermetic/test_test.dart:78]: This test makes sense. Can you also add a test that verifies that `--total-shards` and `--shard-index` are NOT piped through to `fakePackageTest` if they're not provided to the `commandRunner`?

File: packages/flutter_tools/lib/src/commands/test.dart
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/commands/test.dart:297]:
```suggestion
        throwToolExit('If you set --shard-index you need to add --total-shards.');
```

File: packages/flutter_tools/lib/src/commands/test.dart
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/commands/test.dart:290]: This block takes a little bit of though to parse. Could we simplify it, like:

```
    if (totalShards != null && shardIndex == null) {
      throwToolExit('If you set --total-shards you need to also set --shard-index.');
    }
    if (shardIndex != null && totalShards == null) {
      throwToolExit('If you set --shard-index you need to also set --total-shards.');
    }
```

**Review by christopherfujino — CHANGES_REQUESTED** [review]

**Comment by christopherfujino** [comment]
@otuzoglu please refrain from approving PRs until you have been granted [commit access](https://github.com/flutter/flutter/wiki/Tree-hygiene#how-to-review-code). Feel free to comment on PRs if you have constructive feedback.

**Comment by christopherfujino** [comment]
@otuzoglu until granted [contributor access](https://github.com/flutter/flutter/wiki/Contributor-access), please refrain from approving 

> @christopherfujino I'm not sure why the build task fails, I enter into the details and I see a lot of green/red boxes but none of them correspond to this PR.. I see this:
> image

Don't worry about the `flutter-build` CI job, as that is by definition not your fault. It will block merging, but there's a github label `waiting for tree to go green` that can be applied to your PR, and a bot will auto-merge it once the tree goes green, provided all other CI builds are green and PR was approved.

File: packages/flutter_tools/lib/src/commands/test.dart
Comment (danielgomezrico) [reply packages/flutter_tools/lib/src/commands/test.dart:297]: Thanks!

Comment (danielgomezrico) [reply packages/flutter_tools/lib/src/commands/test.dart]: I commited your suggestion and it looks like it uses your user for the commit, do you have a CLI? :P

**Comment by danielgomezrico** [comment]
@christopherfujino I merged your suggestion and it looks like you have not signed the CLA, should I uncommit and push it without your name on that? :P

**Comment by christopherfujino** [comment]
> @christopherfujino I merged your suggestion and it looks like you have not signed the CLA, should I uncommit and push it without your name on that? :P

I have signed the CLA, but this is a bug with the GitHub suggest feature (that I had hoped they fixed) where the merged suggestions get committed as "GitHub bot" or some such.

But yes, the best solution would probably be to `git reset HEAD~N` where `N` is the number of commits from the tip the earliest "bad" commit is, and re-commit them as your user.

**Comment by christopherfujino** [comment]
I overrode the cla-bot because all authors have indeed signed the CLA.

**Comment by danielgomezrico** [comment]
@christopherfujino I made all of the changes, is there any other requested change to merge it? should I click on "resolve conversation"?

File: packages/flutter_tools/test/commands.shard/hermetic/test_test.dart
Comment (christopherfujino) [inline packages/flutter_tools/test/commands.shard/hermetic/test_test.dart:99]: one last nit, could you make these patterns not include the hard-coded numbers, i.e. `isNot(contains('--total-shards'))`. After that LGTM, and we can ship this :)

Comment (danielgomezrico) [reply packages/flutter_tools/test/commands.shard/hermetic/test_test.dart]: done 👍🏻

**Review by christopherfujino — APPROVED** [review]
LGTM

**Comment by hedi-ghodhbane** [comment]
Test coverage percentage is not the same when you run test with these arguments and when you run it without these arguments. Actually when i did `flutter test --total-shards 3 --shard-index 0 --coverage` the test coverage was less then the result with this command `flutter test --coverage`

**Comment by danielgomezrico** [comment]
@hedi-ghodhbane I suggest that you should open a new issue for this :)

**Comment by prolificcoder** [comment]
@danielgomezrico can you give a full example of the command line? Say I want to create four shards, is this how I am supposed to do?

```
flutter test test/folder --total-shards 4 --shard-index  0  test/folder/subfolder1 \
flutter test test/folder --total-shards 4 --shard-index  1  test/folder/subfolder2 \
flutter test test/folder --total-shards 4 --shard-index  2  test/folder/subfolder3 \
flutter test test/folder --total-shards 4 --shard-index  3  test/folder/subfolder4 

```

Am I invoking seperate flutter test commands or just one command? 

Additionally, I want to try to use this with gitlab [parallel](https://docs.gitlab.com/ee/ci/yaml/#parallel) is that possible? (or a github too, I can translate)

**Comment by danielgomezrico** [comment]
@prolificcoder Im not sure about that `\` command, but if you run in one place 
```console
flutter test test/folder --total-shards 4 --shard-index  0  test/folder/subfolder1
```

it will run an small set of the tests and then: 

```console
flutter test test/folder --total-shards 4 --shard-index  1  test/folder/subfolder1
```

It will run another set of the test suite and so on.

Also im not sure about your parallel question, feel free to comment here if you find something else

**Pull request merged**
